test/test_torch.py

```diff
@@ -1909,6 +1909,41 @@ def _test_diagonal(self, dtype, device):
     def test_diagonal(self):
         self._test_diagonal(self, dtype=torch.float32, device='cpu')
 
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_diagonal_multidim(self):
+        x = torch.randn(10, 11, 12, 13)
+        xn = x.numpy()
+        result = torch.diagonal(x, 2, 2, 3)
+        expected = xn.diagonal(2, 2, 3)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
+        result = torch.diagonal(x, 2)
+        expected = torch.diagonal(x, 2, 0, 1)
+        self.assertEqual(expected, result)
+        result = torch.diagonal(x, -2, 1, 2)
+        expected = xn.diagonal(-2, 1, 2)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
+        result = torch.diagonal(x, 0, -2, -1)
+        expected = xn.diagonal(0, -2, -1)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
+        # test non-continguous
+        xp = x.permute(1, 2, 3, 0)
+        result = torch.diagonal(xp, 0, -2, -1)
+        expected = xp.numpy().diagonal(0, -2, -1)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
+        # test that the backward requires grad
+        # we do this is because diagonal_backward uses inplace
+        # operations and gradgradcheck does not catch whether
+        # they works as expected
```

Comment:
Why isn't `gradgradcheck` enough? I don't really understand the comment, but `test_torch` shouldn't check anything related to differentiation

Reply:
Because gradgradcheck doesn't complain when the derivative does not require grad. So when I used the diagonal in the backward, it passed gradgradcheck, but the gradient was marked as not needing a derivative, and double backward was zero.
The test is a bit strange, can I move it to test_autograd.py then?

Reply:
Well usually you just want this for all arguments, and it would improve the test coverage significantly, so I think it's worth it (not as part of this PR).

Reply:
OK, should I drop the check for now in anticipation of the increased test coverage or keep it with a comment?